CMAA JWT Validation Sample

olkroshk opened the pull request:
## Purpose
Adding CMAA JWT validation sample.

## Does this introduce a breaking change?
```
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[x] Documentation content changes
[x] Other... Please describe: code sample for JWT validation
```

## How to Test
* Get the code
* Follow the steps https://github.com/olkroshk/microsoft-azure-attestation-2/tree/dev/olkroshk/samples-update-draft/cmaa.sevsnp.attest.sample#prerequisites

## What to Check
- Sample runs
- With the current status, the validation steps besides `VerifyHostDataClaim` and `VerifyReportDataClaim` for the given sample data must succeed.

## Other Information
n/a

olkroshk committed:
added draft for the sev snp sample

olkroshk committed:
updated readme.md

olkroshk committed:
updated the files

olkroshk committed:
updated readme and the project - the pjrect builds without any warnings

olkroshk committed:
updated the program

olkroshk committed:
updated readme and the program

olkroshk committed:
added verification steps

deschuma commented on cmaa.sevsnp.attest.sample/Program.cs:
```diff
@@ -0,0 +1,422 @@
+    private static async Task<bool> ValidateJwtAsync(string token, string attestationInstanceURL)
+    {
+        // TODO: this is needed to temporary ignore the results. 
+        bool validationSucceeded = true;
```
Is this still needed?

olkroshk replied:
Same as `VerifyTokenSignatureManually`. Do you think we should fail on the first error, or should we continue to see all validation errors?

olkroshk replied:
removed

deschuma commented on cmaa.sevsnp.attest.sample/Program.cs:
```diff
@@ -0,0 +1,422 @@
+            
+            //var verifyTokenSignatureManuallyResult = VerifyTokenSignatureManually(jwtToken, certificates[0]);
+            //if (!verifyTokenSignatureManuallyResult || !verifyTokenSignatureResult)
+            var verifyTokenSignatureResult = VerifyTokenSignature(jwtToken, certificates[0]);
```
Can this be changed to verify the jwt against all the certificates in the /certs response, not just the first one?

olkroshk replied:
Sure, I will update it. I thought validating the first one was sufficient, as we do here: https://github.com/Azure-Samples/microsoft-azure-attestation/blob/5d34c177f9712f821a82ad7799050bedd7850bae/sgx.attest.sample.oe.sdk/validatequotes.net/Helpers/JwtValidationHelper.cs#L24

deschuma replied:
It is usually sufficient but there are weird edge cases where it won't be during BCDR. Not really a problem for a sample itself but since we are sharing with customers would be best to show them the best practice, which is to check all the keys.

deschuma commented on cmaa.sevsnp.attest.sample/Program.cs:
```diff
@@ -0,0 +1,422 @@
+        }
+    }
+
+    private static bool VerifyTokenSignatureManually(JwtSecurityToken jwtToken, X509Certificate2 certificate)
```
Is this still needed?

olkroshk replied:
not really, I left it for the review to decide which method to go with

olkroshk replied:
I'll remove the manual method

olkroshk committed:
- remvoed some temporary methods and variables
- added HOST_DATA value back
- reformatting
- contains some temporary not verified changes

olkroshk committed:
updated logs

gkostal commented on cmaa.sevsnp.attest.sample/AttestationClient.cs:
```diff
@@ -0,0 +1,76 @@
+using Newtonsoft.Json;
+using System.Text;
+
+public class AttestationClient
```
Should this code use the MAA .NET SDK? The attest API called in MAA doesn't have to match the attestation evidence type included in the MAA certificate.

gkostal commented:
I'm not understanding the rationale for the overall factoring of this sample app.  There's already a sample app called "maa.jwt.verifier" that verifies a MAA JWT, including the backing attestation evidence for the MAA service when it's hosted in a SGX TEE.  Why isn't the approach here to enlighten the "maa.jwt.verifier" app so that it also knows how to verify C-ACI attestation evidence?

gkostal commented:
I think any verification sample code we share should factor out in isolation exactly what is trusted by default by the application (e.g., PRSS signing key captured as MRSIGNER for SGX, PRSS signing key that's used in CCE policy that's then hashed to match HOSTDATA for C-ACI, etc.).

We want to educate the customer on specifically what their required trust anchors are, and then the code leverages these trust anchors to validate the (C)MAA attestation evidence embedded in the signing certificate.

gkostal commented:
See slide 12 in [this](https://microsoft.sharepoint.com/:p:/r/teams/AzureSecurityCompliance/Security/_layouts/15/Doc.aspx?sourcedoc=%7BC2FDC695-33AB-43A0-B62E-86257323E305%7D&file=[Maa.Container.Migration.Status.Jan.2024.pptx](https://microsoft.sharepoint.com/:p:/r/teams/AzureSecurityCompliance/Security/_layouts/15/Doc.aspx?sourcedoc=%7BC2FDC695-33AB-43A0-B62E-86257323E305%7D&file=Maa.Container.Migration.Status.Jan.2024.pptx&action=edit&mobileredirect=true&share=IQGVxv3CqzOgQ7YuhiVzI-MFASfABtpsOUImi2Qq8Bkle5Y)&action=edit&mobileredirect=true&share=IQGVxv3CqzOgQ7YuhiVzI-MFASfABtpsOUImi2Qq8Bkle5Y) PPT.  The verification code in the sample app should a) start from two trust anchors expressed as keys in a separate file/class, and b) go through all the steps outlined in the slide including verifying the UVM COSE sign1 document and verifying that its embedded launch measurement matches the one found in the embedded SEVSNP report in the certificate extension.  Ideally the documentation tells the reader where to find the trusted PRSS signing keys on an Azure MAA web site.

![image](https://github.com/user-attachments/assets/36ee4c72-2210-4e87-9696-43b9dec038d0)
